### DISCUSSION TITLE
Child boxes permissions

### DISCUSSION THREAD

**Discussion by gldrenthe89** [OFFLINE-GmbH/oc-boxes-support#145] [Ideas and feature requests]
We want to limit the customer to add child boxes to a box. Can this be added as a permission?

**Comment by tobias-kuendig** [comment]
@gldrenthe89 this is currently not supported and would need an addition to the editor UI. But the requirement is rather specific, so I don't think it is worth it to generally add it to the plugin.

For a pragmatic solution to this problem: you might be able to hide relevant UI elements using some injected CSS and prevent the user from creating content this way.

**Reply by gldrenthe89** [reply]
Maybe not a direct permission then. But a event? like when the user wants to add a child box we can hook into the event and set the optional child box list to an empty list? 

Injecting CSS is not something we would prefer in the backend

**Reply by tobias-kuendig** [reply]
Looking into this again, an event to filter the partials when the selector opens is not viable as well, since the editor gets all its state on page load and then does no further communication with the backend in that regard.

But what you could do is filter the partials using the `FILTER_PARTIALS` event. There you can unset the allowed `children` property for any given partial.

This will prevent the user to add a child for any Box instance with that partial. You cannot decide on a box-by-box basis, but depending on your use-case, this is not requried.

```php
    // app/Provider.php
    public function boot()
    {
        parent::boot();

        \Event::listen(\OFFLINE\Boxes\Classes\Events::FILTER_PARTIALS, function ($partials) {
            $canAddChildren = false; // Set this depending on the logged in user.

            if ( ! $canAddChildren) {
                $partials->get('example')->config->children = []; // No longer allow children.
            }

            return $partials;
        });
    }
```

**Reply by gldrenthe89** [reply]
> Looking into this again, an event to filter the partials when the selector opens is not viable as well, since the editor gets all its state on page load and then does no further communication with the backend in that regard.
> 
> But what you could do is filter the partials using the `FILTER_PARTIALS` event. There you can unset the allowed `children` property for any given partial.
> 
> This will prevent the user to add a child for any Box instance with that partial. You cannot decide on a box-by-box basis, but depending on your use-case, this is not requried.
> 
> ```
>     // app/Provider.php
>     public function boot()
>     {
>         parent::boot();
> 
>         \Event::listen(\OFFLINE\Boxes\Classes\Events::FILTER_PARTIALS, function ($partials) {
>             $canAddChildren = false; // Set this depending on the logged in user.
> 
>             if ( ! $canAddChildren) {
>                 $partials->get('example')->config->children = []; // No longer allow children.
>             }
> 
>             return $partials;
>         });
>     }
> ```

Thanks for the information. We will take this into considiration.
